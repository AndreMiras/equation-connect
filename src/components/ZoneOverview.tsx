import { FC, ReactChild, ReactChildren, useEffect, useState } from "react";
import { useHref } from "react-router-dom";
import { Accordion, ListGroup } from "react-bootstrap";
import {
  DeviceStatus,
  ZoneOverviewType,
  setZonePowerOff,
  setZonePreset,
} from "equation-connect";
import Preset from "./Preset";

interface ListGroupItemProps {
  children: ReactChild | ReactChildren;
  to: string;
}
const ListGroupItem: FC<ListGroupItemProps> = ({ children, to }) => {
  const href = useHref(to);
  return (
    <ListGroup.Item action href={href}>
      {children}
    </ListGroup.Item>
  );
};

interface DeviceListProps {
  devices: string[];
}
const DeviceList: FC<DeviceListProps> = ({ devices }) => (
  <ListGroup>
    {devices.map((device) => (
      <ListGroupItem key={device} to={`/devices/${device}`}>
        {device}
      </ListGroupItem>
    ))}
  </ListGroup>
);

interface ZoneOverviewProps {
  installationId: string;
  zone: ZoneOverviewType;
}
const ZoneOverview: FC<ZoneOverviewProps> = ({ installationId, zone }) => {
  const [statusState, setStatusState] = useState(DeviceStatus.Comfort);
  const [powerState, setPowerState] = useState(false);
  const { id, name, type, comfort, eco, temp, power, mode, status } = zone;
  const devices = Object.keys(zone.devices || {});
  const onPreset = (newStatus: DeviceStatus) => {
    setZonePreset(installationId, id!, newStatus);
    setStatusState(newStatus);
    setPowerState(true);
  };
  const onPowerOff = () => {
    setZonePowerOff(installationId, id!);
    setPowerState(false);
  };
  useEffect(() => {
    setPowerState(power);
    setStatusState(status);
  }, [power, status]);
  return (
    <Accordion className="mb-3" defaultActiveKey="0">
      <Accordion.Item eventKey="0">
        <Accordion.Header>{name}</Accordion.Header>
        <Accordion.Body>
          <ListGroup>
            <DeviceList devices={devices} />
          </ListGroup>
          <Preset
            name={`options-${id}`}
            status={statusState}
            onPreset={onPreset}
            power={powerState}
            onPowerOff={onPowerOff}
          />
        </Accordion.Body>
      </Accordion.Item>
      <Accordion.Item eventKey="1">
        <Accordion.Header>Advanced</Accordion.Header>
        <Accordion.Body>
          <ul>
            <li>id: {id}</li>
            <li>temp: {temp}&deg;</li>
            <li>type: {type}</li>
            <li>comfort: {comfort}&deg;</li>
            <li>eco: {eco}&deg;</li>
            <li>temp: {temp}&deg;</li>
            <li>power: {power.toString()}</li>
            <li>mode: {mode}</li>
          </ul>
          <Preset
            name={`options-${id}`}
            status={statusState}
            onPreset={onPreset}
            power={powerState}
            onPowerOff={onPowerOff}
          />
        </Accordion.Body>
      </Accordion.Item>
      <Accordion.Item eventKey="2">
        <Accordion.Header>Debug</Accordion.Header>
        <Accordion.Body>
          <pre>{JSON.stringify(zone, null, 2)}</pre>
        </Accordion.Body>
      </Accordion.Item>
    </Accordion>
  );
};

export default ZoneOverview;
