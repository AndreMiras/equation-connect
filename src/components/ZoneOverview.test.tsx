import React from "react";
import { HashRouter as Router } from "react-router-dom";
import { render } from "@testing-library/react";
import ZoneOverview from "./ZoneOverview";

const zoneProps = {
  comfort: 22,
  devices: {
    D89DDF3A7D80D89DDF3A7D80: true,
    F9097F334FC4F9097F334FC4: true,
  },
  eco: 16,
  id: "-MpGOkgR4zExLBUDUafq",
  mode: "manual",
  name: "Living room",
  power: true,
  temp: 21,
  type: "radiator",
};

test("ZoneOverview renders correctly", () => {
  const { asFragment } = render(
    <Router>
      <ZoneOverview zone={zoneProps} />
    </Router>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("ZoneOverview devices is optional", () => {
  const zoneProps2 = { ...zoneProps };
  delete zoneProps2.devices;
  const { asFragment } = render(
    <Router>
      <ZoneOverview zone={zoneProps2} />
    </Router>
  );
  expect(asFragment()).toMatchSnapshot();
});
