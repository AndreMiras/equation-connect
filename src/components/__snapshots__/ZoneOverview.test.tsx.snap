// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZoneOverview devices is optional 1`] = `
<DocumentFragment>
  <div
    class="mb-3 accordion"
  >
    <div
      class="accordion-item"
    >
      <h2
        class="accordion-header"
      >
        <button
          aria-expanded="true"
          class="accordion-button"
          type="button"
        >
          Living room
        </button>
      </h2>
      <div
        class="accordion-collapse collapse show"
      >
        <div
          class="accordion-body"
        >
          <ul>
            <li>
              id: -MpGOkgR4zExLBUDUafq
            </li>
            <li>
              temp: 21°
            </li>
            <li>
              type: radiator
            </li>
            <li>
              comfort: 22°
            </li>
            <li>
              eco: 16°
            </li>
            <li>
              temp: 21°
            </li>
            <li>
              power: true
            </li>
            <li>
              mode: manual
            </li>
            <li>
              devices: 0
            </li>
            <ul />
          </ul>
        </div>
      </div>
    </div>
    <div
      class="accordion-item"
    >
      <h2
        class="accordion-header"
      >
        <button
          aria-expanded="false"
          class="accordion-button collapsed"
          type="button"
        >
          Debug
        </button>
      </h2>
      <div
        class="accordion-collapse collapse"
      >
        <div
          class="accordion-body"
        >
          <pre>
            {
  "comfort": 22,
  "eco": 16,
  "id": "-MpGOkgR4zExLBUDUafq",
  "mode": "manual",
  "name": "Living room",
  "power": true,
  "temp": 21,
  "type": "radiator"
}
          </pre>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ZoneOverview renders correctly 1`] = `
<DocumentFragment>
  <div
    class="mb-3 accordion"
  >
    <div
      class="accordion-item"
    >
      <h2
        class="accordion-header"
      >
        <button
          aria-expanded="true"
          class="accordion-button"
          type="button"
        >
          Living room
        </button>
      </h2>
      <div
        class="accordion-collapse collapse show"
      >
        <div
          class="accordion-body"
        >
          <ul>
            <li>
              id: -MpGOkgR4zExLBUDUafq
            </li>
            <li>
              temp: 21°
            </li>
            <li>
              type: radiator
            </li>
            <li>
              comfort: 22°
            </li>
            <li>
              eco: 16°
            </li>
            <li>
              temp: 21°
            </li>
            <li>
              power: true
            </li>
            <li>
              mode: manual
            </li>
            <li>
              devices: 2
            </li>
            <ul>
              <li>
                <a
                  href="#/devices/D89DDF3A7D80D89DDF3A7D80"
                >
                  D89DDF3A7D80D89DDF3A7D80
                </a>
              </li>
              <li>
                <a
                  href="#/devices/F9097F334FC4F9097F334FC4"
                >
                  F9097F334FC4F9097F334FC4
                </a>
              </li>
            </ul>
          </ul>
        </div>
      </div>
    </div>
    <div
      class="accordion-item"
    >
      <h2
        class="accordion-header"
      >
        <button
          aria-expanded="false"
          class="accordion-button collapsed"
          type="button"
        >
          Debug
        </button>
      </h2>
      <div
        class="accordion-collapse collapse"
      >
        <div
          class="accordion-body"
        >
          <pre>
            {
  "comfort": 22,
  "devices": {
    "D89DDF3A7D80D89DDF3A7D80": true,
    "F9097F334FC4F9097F334FC4": true
  },
  "eco": 16,
  "id": "-MpGOkgR4zExLBUDUafq",
  "mode": "manual",
  "name": "Living room",
  "power": true,
  "temp": 21,
  "type": "radiator"
}
          </pre>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
